### ISSUE TITLE
`quickstart.sh` is 2.5x slower with tools v5.2

### ISSUE THREAD

**Issue by zstadler** [openmaptiles/openmaptiles#967]
`quickstart.sh` is 2.5 times slower after "Use tools v5.2" #903 

It looks like the slowdown is caused primarily by `import-sql` which runs for 81 minutes instead of 2 minutes.

| commit | `quickstart.sh` time | `import-sql` time
|---------|------:|---:|
| c6fff87 | 3116s | 121s |
| 907fc58 | 7740s | 4867s |
| **Increase** | **4646s** | **4744s** |

Note: c6fff87 is the parent of 907fc58

**Comment by zstadler** [comment]
Strangely, there is no difference between the contents of the `build/sql` directory of the two versions.

**Comment by TomPohys** [comment]
I cannot find anything that changed import-sql. For what area do you test the times?

**Comment by zstadler** [comment]
I used the israel-and-palestine area from Geofabrik

**Comment by zstadler** [comment]
Unfortunately, in Estonia things are worse than in Israel-and-Palestine

hash | quickstart.sh time | import-sql time
-- | --: | --:
907fc58 | 8743s | 8043s
c6fff87 | 682s | 122s
**Increase** | **8061s**| **7921s**
slowdown | 12.8 | 65.9

In Iceland, on the other hand, is showing a relatively small slowdown


hash | quickstart.sh time | import-sql time
-- | --: | --:
907fc58 | 682s | 178s
c6fff87 | 574s | 103s
**Increase** | **108s** | **75s**
slowdown | 1.2 | 1.7

**Comment by frodrigo** [comment]
Now there is 5 SQL runner in parallel, it may affect the loading time on your setup. Check at `.env` `MAX_PARALLEL_PSQL=5`.

**Comment by zstadler** [comment]
Both versions and both measurements are using 5 SQL runner in parallel. Only this number is not configurable by `.env` in c6fff87.

**Comment by zstadler** [comment]
A locally-built Docker image of `openmaptiles-tools` version 5.1 shows the same bad performance as the public docker image of 5.2. The public docker image of 5.1 consistently shows the good performance.

How can that be?

**Comment by zstadler** [comment]
This is the longest operation in `import-sql` is the [creation of the `osm_transportation_name_linestring` materialized view](https://github.com/openmaptiles/openmaptiles/blob/194b2fbb701c4739ad0c58e1d4fe72728e4e0cf9/layers/transportation_name/update_transportation_name.sql#L40-L79) which takes 80 minutes.

The second-longest operation is [`SELECT update_osm_poi_point();`](https://github.com/openmaptiles/openmaptiles/blob/194b2fbb701c4739ad0c58e1d4fe72728e4e0cf9/layers/poi/update_poi_point.sql#L25) which takes 50 minutes. [`SELECT update_poi_polygon(true);`](https://github.com/openmaptiles/openmaptiles/blob/194b2fbb701c4739ad0c58e1d4fe72728e4e0cf9/layers/poi/update_poi_polygon.sql#L46) takes another 12 minutes.

Perhaps the solution to this issue is in optimizing these bottlenecks.

@frodrigo , 
Do you think this would be covered by #814?

**Comment by nyurik** [comment]
Note that if this is related to index regeneration concurrency, than it might depend on the type of harddrive -- concurrent generation might improve speed with SSDs while decrease speed with the spinning rust ones, especially if the machine has low RAM.

**Comment by frodrigo** [comment]
> @frodrigo ,
> Do you think this would be covered by #814?

No, it is linked to the update, not the initial import.

**Comment by hanchao** [comment]
Worse in taiwan

![image](https://user-images.githubusercontent.com/1942091/107014758-01c13180-67d7-11eb-8c70-1458ce221148.png)

**Comment by tbicr** [comment]
Same issue for me with `quickstart.sh belarus`call, look like some triggers issue there:

```
openmaptiles=# select count(*) from osm_poi_point
                           
    WHERE station = 'subway'
      AND subclass = 'station';
 count 
-------
     0
(1 row)

Time: 69.185 ms


openmaptiles=# explain analyze  UPDATE osm_poi_point
    SET subclass = 'subway'
    WHERE station = 'subway'
      AND subclass = 'station';
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Update on osm_poi_point  (cost=0.00..18079.07 rows=1 width=245) (actual time=86.576..86.576 rows=0 loops=1)
   ->  Seq Scan on osm_poi_point  (cost=0.00..18079.07 rows=1 width=245) (actual time=86.569..86.569 rows=0 loops=1)
         Filter: (((station)::text = 'subway'::text) AND ((subclass)::text = 'station'::text))
         Rows Removed by Filter: 122138
 Planning time: 0.052 ms
 Trigger trigger_flag: time=0.153 calls=1
 Execution time: 86.754 ms
(7 rows)

Time: 7924.357 ms


openmaptiles=# begin;
BEGIN
Time: 0.057 ms
openmaptiles=# drop trigger trigger_flag on osm_poi_point ;
DROP TRIGGER
Time: 1001.189 ms
openmaptiles=# explain analyze  UPDATE osm_poi_point
    SET subclass = 'subway'
    WHERE station = 'subway'
      AND subclass = 'station';
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Update on osm_poi_point  (cost=0.00..18079.07 rows=1 width=245) (actual time=79.448..79.448 rows=0 loops=1)
   ->  Seq Scan on osm_poi_point  (cost=0.00..18079.07 rows=1 width=245) (actual time=79.446..79.446 rows=0 loops=1)
         Filter: (((station)::text = 'subway'::text) AND ((subclass)::text = 'station'::text))
         Rows Removed by Filter: 122138
 Planning time: 0.297 ms
 Execution time: 79.514 ms
(6 rows)

Time: 80.389 ms
openmaptiles=# rollback;


openmaptiles=# \d osm_poi_point
                                  Table "public.osm_poi_point"
   Column    |         Type         |                         Modifiers                          
-------------+----------------------+------------------------------------------------------------
 id          | integer              | not null default nextval('osm_poi_point_id_seq'::regclass)
 osm_id      | bigint               | 
 name        | character varying    | 
 name_en     | character varying    | 
 name_de     | character varying    | 
 tags        | hstore               | 
 subclass    | character varying    | 
 mapping_key | character varying    | 
 station     | character varying    | 
 funicular   | character varying    | 
 information | character varying    | 
 uic_ref     | character varying    | 
 religion    | character varying    | 
 level       | integer              | 
 indoor      | boolean              | 
 layer       | integer              | 
 sport       | character varying    | 
 geometry    | geometry(Point,3857) | 
 agg_stop    | integer              | 
Indexes:
    "osm_poi_point_pkey" PRIMARY KEY, btree (id)
    "osm_poi_point_geom" gist (geometry)
    "osm_poi_point_osm_id_idx" btree (osm_id)
Triggers:
    trigger_flag AFTER INSERT OR DELETE OR UPDATE ON osm_poi_point FOR EACH STATEMENT EXECUTE PROCEDURE poi_point.flag()


CREATE OR REPLACE FUNCTION poi_point.flag() RETURNS trigger AS
$$
BEGIN
    INSERT INTO poi_point.updates(t) VALUES ('y') ON CONFLICT(t) DO NOTHING;
    RETURN NULL;
END;
$$ LANGUAGE plpgsql;


openmaptiles=# select count(*) from poi_point.updates;
 count 
-------
     0
(1 row)

Time: 0.258 ms


openmaptiles=# \d poi_point.updates
                             Table "poi_point.updates"
 Column |  Type   |                           Modifiers                            
--------+---------+----------------------------------------------------------------
 id     | integer | not null default nextval('poi_point.updates_id_seq'::regclass)
 t      | text    | 
Indexes:
    "updates_pkey" PRIMARY KEY, btree (id)
    "updates_t_key" UNIQUE CONSTRAINT, btree (t)
Triggers:
    trigger_refresh AFTER INSERT ON poi_point.updates DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE poi_point.refresh()


CREATE OR REPLACE FUNCTION poi_point.refresh() RETURNS trigger AS
$$
DECLARE
    t TIMESTAMP WITH TIME ZONE := clock_timestamp();
BEGIN
    RAISE LOG 'Refresh poi_point';
    PERFORM update_osm_poi_point();
    REFRESH MATERIALIZED VIEW osm_poi_stop_centroid;
    REFRESH MATERIALIZED VIEW osm_poi_stop_rank;
    PERFORM update_osm_poi_point_agg();
    -- noinspection SqlWithoutWhere
    DELETE FROM poi_point.updates;

    RAISE LOG 'Refresh poi_point done in %', age(clock_timestamp(), t);
    RETURN NULL;
END;
$$ LANGUAGE plpgsql;
```

**Comment by tbicr** [comment]
My resent investigation give next highlights:
`update_tags` slow, that reasoned by `get_basic_names` slowness, that reasoned by `get_latin_name` slowness, that reasoned by `osml10n_get_name_without_brackets_from_tags` slowness, `osml10n_get_name_without_brackets_from_tags` defined in `osml10n` extension.

**Comment by frodrigo** [comment]
The issue on mapnik-german-l10n is already fixed, but the fix not released
https://github.com/giggls/mapnik-german-l10n/issues/54

I add this on my side, waiting to mapnik-german-l10n release
https://github.com/makina-maps/openmaptiles/commit/5b69265c626cd744766089ba625938f7f8daf41d

**Comment by tbicr** [comment]
yeah, find that `osml10n_get_country` super slow there as it have issue with index usage: 

```
# EXPLAIN (ANALYZE, VERBOSE, COSTS,BUFFERS, TIMING) select (SELECT country_code  from country_osm_grid g where st_contains(g.geometry, st_centroid(st_transform(p.geometry,4326))) order by area limit 1) as x from osm_poi_polygon p limit 100; 
                                                                        QUERY PLAN                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.00..1134066.87 rows=100 width=12) (actual time=120.314..12198.337 rows=100 loops=1)
   Output: ((SubPlan 1))
   Buffers: shared hit=1288817
   ->  Seq Scan on public.osm_poi_polygon p  (cost=0.00..805278204.32 rows=71008 width=12) (actual time=120.313..12198.309 rows=100 loops=1)
         Output: (SubPlan 1)
         Buffers: shared hit=1288817
         SubPlan 1
           ->  Limit  (cost=11340.59..11340.59 rows=1 width=11) (actual time=121.980..121.980 rows=1 loops=100)
                 Output: g.country_code, g.area
                 Buffers: shared hit=1288811
                 ->  Sort  (cost=11340.59..11359.94 rows=7740 width=11) (actual time=121.979..121.979 rows=1 loops=100)
                       Output: g.country_code, g.area
                       Sort Key: g.area
                       Sort Method: quicksort  Memory: 25kB
                       Buffers: shared hit=1288811
                       ->  Seq Scan on public.country_osm_grid g  (cost=0.00..11301.89 rows=7740 width=11) (actual time=15.901..121.970 rows=1 loops=100)
                             Output: g.country_code, g.area
                             Filter: st_contains(g.geometry, st_centroid(st_transform(p.geometry, 4326)))
                             Rows Removed by Filter: 23218
                             Buffers: shared hit=1288811
 Planning time: 0.069 ms
 Execution time: 12198.370 ms
(22 rows)

Time: 12198.713 ms
```

so removing it really helps, thanks.

**Comment by tbicr** [comment]
add temp changes to makefile decrease run of `import-sql` from hours to 10 min for `quickstart.sh belarus`:

```.PHONY: import-sql
import-sql: all start-db-nowait
	$(DOCKER_COMPOSE) run $(DC_OPTS) openmaptiles-tools psql.sh -v ON_ERROR_STOP=1 -P pager=off \
	    -c "CREATE or REPLACE FUNCTION osml10n_geo_translit(name text, place geometry DEFAULT NULL) RETURNS TEXT AS ' BEGIN IF (place IS NULL) THEN return osml10n_cc_transscript(name,''aq''); ELSE return(osml10n_cc_transscript(name,NULL)); END IF; END; ' LANGUAGE plpgsql STABLE;"
	$(DOCKER_COMPOSE) run $(DC_OPTS) openmaptiles-tools sh -c 'pgwait && import-sql' | \
	  awk -v s=": WARNING:" '1{print; fflush()} $$0~s{print "\n*** WARNING detected, aborting"; exit(1)}'```

**Comment by maxammann** [comment]
I successfully rendered level 0-14 with @tbicr's commit. Initially I stopped the `import-sql` command after 17 days without progress.

With that commit the whole generation process only took 5 days. Can anyone highlight what the implications of this workaround are?
